modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/statistic/SqlStatisticManagerImpl.java

```diff
@@ -0,0 +1,122 @@
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * Statistic manager. Provide and manage update of statistics for SQL.
+ */
+public class SqlStatisticManagerImpl implements SqlStatisticManager {
+    private static final IgniteLogger LOG = Loggers.forClass(SqlStatisticManagerImpl.class);
+    private static final long DEFAULT_TABLE_SIZE = 1_000_000;
+    private static final long MINIMUM_TABLE_SIZE = 1_000L;
+    private static final Pair<Long, Long> DEFAULT_VALUE = new Pair<>(DEFAULT_TABLE_SIZE, 0L);
+
+    private final TableManager tableManager;
+
+    private final ConcurrentMap<Integer, Pair<Long, Long>> tableSizeMap = new ConcurrentHashMap<>();
+
+    private volatile long thresholdTimeToPostponeUpdateMs = TimeUnit.MINUTES.toMillis(1);
+
+    /** Constructor. */
+    public SqlStatisticManagerImpl(TableManager tableManager, CatalogManager catalogManager) {
+        this.tableManager = tableManager;
+
+        catalogManager.listen(CatalogEvent.TABLE_DROP, fromConsumer(event -> {
+            int tableId = ((DropTableEventParameters) event).tableId();
+            onTableDrop(tableId);
+        }));
+    }
+
+    /**
+     * Returns approximate number of rows in table by their id. Returns the previous known value or
+     * {@value SqlStatisticManagerImpl#DEFAULT_TABLE_SIZE} as default value. Can start process to update asked statistics in background to
+     * have updated values for future requests.
+     */
+    @Override
+    public long tableSize(int tableId) {
+        updateTableSizeStatistics(tableId);
+        long ts = tableSizeMap.getOrDefault(tableId, DEFAULT_VALUE).getFirst();
+
+        return Math.max(ts, MINIMUM_TABLE_SIZE);
+    }
+
+    /** Update table size statistic in the background if it required. */
+    private void updateTableSizeStatistics(int tableId) {
+        tableSizeMap.putIfAbsent(tableId, DEFAULT_VALUE);
+        Pair<Long, Long> pair = tableSizeMap.get(tableId);
+        long currTs = FastTimestamps.coarseCurrentTimeMillis();
+        long lastUpdateTime = pair.getSecond();
+
+        if (lastUpdateTime <= currTs - thresholdTimeToPostponeUpdateMs) {
+            // Prevent to run update for the same table twice concurrently.
+            if (!tableSizeMap.replace(tableId, pair, new Pair<>(pair.getFirst(), currTs))) {
+                return;
+            }
+
+            TableViewInternal tableView = tableManager.cachedTable(tableId);
+            if (tableView == null) {
+                LOG.debug("There is no table to update statistics [id={}]", tableId);
+                return;
+            }
+
+            // just request new table size in background.
+            tableView.internalTable().estimatedSize()
+                    .thenApply(size -> {
+                        // the table can be concurrently dropped and we shouldn't put new value in this case.
+                        tableSizeMap.computeIfPresent(tableId, (k, v) -> new Pair<>(size, currTs));
```

Comment:
does it make sense that value has not been updated concurrently?

Reply:
Yes, it is pointless to do this work twice simultaneously.